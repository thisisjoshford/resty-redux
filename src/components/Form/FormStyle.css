.Form {
  display: flex;
  justify-content: left;
 
  label {
    cursor: pointer;
    margin-left: .5rem;
  }

  input[type="radio"]:hover {
    background-color: purple;
    color: white;

  }
  input[type="radio"]:checked+label {
    background-color: gold
  }
}

.Url {
  height: 20px;
  width: 300px;
  box-shadow: grey 3px 3px;
}

.Radio {
  font-size: x-large;
}
