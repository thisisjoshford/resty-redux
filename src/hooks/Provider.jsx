import React, { createContext, useState, useContext } from 'react';
import { fetchRequest } from '../services/fetchRequest';

const RestyContext = createContext();

export const RestyProvider = ({ children }) => {
  const [url, setUrl] = useState('');
  const [method, setMethod] = useState('GET');
  const [textBody, setTextBody] = useState('');
  const [response, setResponse] = useState([]);
  const [history, setHistory] = useState([]);

  const onChange = ({ target }) => {
    if(target.name === 'url') setUrl(target.value);
    if(target.name === 'method') setMethod(target.value);
    if(target.name === 'textBody') setTextBody(target.value);
  };

  const onSubmit = event => {
    event.preventDefault();
    fetchRequest(url, method, textBody)
      .then(res => {
        setResponse(res);
        setHistory(prevHistory => 
          [{ url, method }, ...prevHistory]);
      });
  };

  return (
    <RestyContext.Provider 
      value ={{ 
        url, 
        method, 
        textBody, 
        response,
        history, 
        onChange, 
        onSubmit 
      }}>
      {children}
    </RestyContext.Provider>
  );
};

export const useProvider = () => {
  const context = useContext(RestyContext);
  return context;
};
