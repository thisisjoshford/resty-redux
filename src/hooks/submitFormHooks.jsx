import { fetchRequest } from '../services/fetchRequest';
import { useDispatch, useSelector } from 'react-redux';
import { getUrl, getMethod, getTextBody, getResponse, getHistory } from '../selectors/selectors';
import { addResponse, addHistory, setUrl, setMethod, setTextBody } from '../actions/restfulActions';

export const useSubmitForm = () => {
  const dispatch = useDispatch();
  const url = useSelector(getUrl);
  const method = useSelector(getMethod);
  const textBody = useSelector(getTextBody);
  const response = useSelector(getResponse);
  const history = useSelector (getHistory);

  const onChange = ({ target }) => {
    if(target.name === 'url') dispatch(setUrl(target.value));
    if(target.name === 'method') dispatch(setMethod(target.value));
    if(target.name === 'textBody') dispatch(setTextBody(target.value));
  };

  const onSubmit = event => {
    event.preventDefault();
    dispatch(fetchRequest(url, method, textBody)
      .then(res => {
        addResponse(res);
        addHistory(prevHistory => 
          [{ url, method }, ...prevHistory]);
      }));
  };

  return {
    url,
    method,
    textBody,
    response,
    history,
    onChange,
    onSubmit
  };
};
