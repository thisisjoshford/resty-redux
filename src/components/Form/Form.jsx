import React from 'react';
import style from './FormStyle.css';
import { useProvider } from '../../hooks/Provider';

const Form = () => {
  const { url, method, textBody, onChange, onSubmit } = useProvider();
  return (
    <div className={style.Form}>
      <form className={style.Form} onSubmit={onSubmit}>
      
        <label>
          <input 
            type="text"
            name="url"
            className={style.Url}
            placeholder='enter target URL... '
            value={url}
            onChange={onChange}
          >
          </input>
        </label>
      
        <div>
          <label className={style.Radio}>
            <input 
              className={style.RadioButton}
              type="radio" 
              name="method" 
              value="GET"
              checked={method === 'GET'} 
              onChange={onChange}
            /> GET 
          </label>
          <label className={style.Radio}>
            <input 
              className={style.RadioButton}
              type="radio" 
              name="method" 
              value="POST"
              checked={method === 'POST'}
              onChange={onChange}
            /> POST 
          </label>
          <label className={style.Radio}>
            <input 
              className={style.RadioButton}
              type="radio" 
              name="method" 
              value="PUT"
              checked={method === 'PUT'}
              onChange={onChange} 
            /> PUT 
          </label>
          <label className={style.Radio}>
            <input 
              className={style.RadioButton}
              type="radio" 
              name="method" 
              value="PATCH" 
              checked={method === 'PATCH'}
              onChange={onChange}
            /> PATCH 
          </label>
          <label className={style.Radio}>
            <input 
              className={style.RadioButton}
              type="radio" 
              name="method" 
              value="DELETE"
              checked={method === 'DELETE'}
              onChange={onChange}
            /> DELETE 
          </label>
        </div>

        <textarea 
          placeholder="JSON Body"
          name="textBody"
          value={textBody}
          onChange={onChange}
        >
        </textarea>

        <button type="submit">GO!</button>

      </form>
    </div>
  );
};

export default Form;
