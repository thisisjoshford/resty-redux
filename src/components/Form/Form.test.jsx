import React from 'react';
import { shallow } from 'enzyme';
import Form from './Form';
import { RestyProvider } from '../../hooks/Provider';

describe('Form component', () => {

  const url = 'http://www.randomapi.com/';

  it('renders Form', () => {
    const wrapper = shallow(
      <RestyProvider>
        <Form />
      </RestyProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
