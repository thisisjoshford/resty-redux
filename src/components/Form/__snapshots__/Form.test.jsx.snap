// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component renders Form 1`] = `
<ContextProvider
  value={
    Object {
      "history": Array [],
      "method": "GET",
      "onChange": [Function],
      "onSubmit": [Function],
      "response": Array [],
      "textBody": "",
      "url": "",
    }
  }
>
  <Form />
</ContextProvider>
`;
